// ml-peak-shape-generator/lib/shapes/1d/pseudoVoigt/PseudoVoigt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getData = exports.getFactor = exports.getArea = exports.fwhmToWidth = exports.widthToFWHM = exports.fct = exports.PseudoVoigt = void 0;
const constants_1 = require("../../../util/constants");
const Shape1D_1 = require("../Shape1D");
const Gaussian_1 = require("../gaussian/Gaussian");
const Lorentzian_1 = require("../lorentzian/Lorentzian");
class PseudoVoigt extends Shape1D_1.Shape1D {
    constructor(options = {}) {
        super();
        const { fwhm = 500, height, mu = 0.5 } = options;
        this.mu = mu;
        this.fwhm = fwhm;
        this.height =
            height === undefined
                ? 1 /
                    ((mu / Math.sqrt(-constants_1.GAUSSIAN_EXP_FACTOR / Math.PI)) * fwhm +
                        ((1 - mu) * fwhm * Math.PI) / 2)
                : height;
    }
    fwhmToWidth(fwhm = this.fwhm, mu = this.mu) {
        return fwhmToWidth(fwhm, mu);
    }
    widthToFWHM(width, mu = this.mu) {
        return widthToFWHM(width, mu);
    }
    fct(x) {
        return fct(x, this.fwhm, this.mu);
    }
    getArea() {
        return getArea({ fwhm: this.fwhm, height: this.height, mu: this.mu });
    }
    getFactor(area) {
        return getFactor(area);
    }
    getData(options = {}) {
        const { length, factor } = options;
        return getData({
            fwhm: this.fwhm,
            height: this.height,
            mu: this.mu,
            factor,
            length,
        });
    }
}
exports.PseudoVoigt = PseudoVoigt;
/**
 * Return a parameterized function of a pseudo voigt shape (see README for equation).
 * @param x - x value to calculate.
 * @param fwhm - full width half maximum
 * @returns - the y value of pseudo voigt with the current parameters.
 */
function fct(x, fwhm, mu) {
    return (1 - mu) * Lorentzian_1.fct(x, fwhm) + mu * Gaussian_1.fct(x, fwhm);
}
exports.fct = fct;
/**
 * Compute the value of Full Width at Half Maximum (FWHM) from the width between the inflection points.
 * @param width - Width between the inflection points
 * @param [mu=0.5] Ratio of gaussian contribution in the shape
 * @returns fwhm
 */
function widthToFWHM(width, mu = 0.5) {
    return width * (mu * constants_1.ROOT_2LN2_MINUS_ONE + 1);
}
exports.widthToFWHM = widthToFWHM;
/**
 * Compute the value of width between the inflection points from Full Width at Half Maximum (FWHM).
 * @param fwhm - Full Width at Half Maximum.
 * @param [mu=0.5] Ratio of gaussian contribution in the shape
 * @returns width
 */
function fwhmToWidth(fwhm, mu = 0.5) {
    return fwhm / (mu * constants_1.ROOT_2LN2_MINUS_ONE + 1);
}
exports.fwhmToWidth = fwhmToWidth;
/**
 * Calculate the area of a specific shape.
 * @returns returns the area of the specific shape and parameters.
 */
function getArea(options) {
    const { fwhm, height = 1, mu = 0.5 } = options;
    if (fwhm === undefined) {
        throw new Error('should pass fwhm or sd parameters');
    }
    return (fwhm * height * (mu * constants_1.ROOT_PI_OVER_LN2 + (1 - mu) * Math.PI)) / 2;
}
exports.getArea = getArea;
/**
 * Calculate the number of times FWHM allows to reach a specific area coverage.
 * @param [area=0.9999] Expected area to be covered.
 * @returns
 */
function getFactor(area = 0.9999, mu = 0.5) {
    return mu < 1 ? Lorentzian_1.getFactor(area) : Gaussian_1.getFactor(area);
}
exports.getFactor = getFactor;
/**
 * Calculate intensity array of a pseudo voigt shape.
 * @returns {Float64Array} y values
 */
function getData(options = {}) {
    let { length, factor = getFactor(), fwhm = 500, height, mu = 0.5 } = options;
    if (!height) {
        height =
            1 /
                ((mu / Math.sqrt(-constants_1.GAUSSIAN_EXP_FACTOR / Math.PI)) * fwhm +
                    ((1 - mu) * fwhm * Math.PI) / 2);
    }
    if (!length) {
        length = Math.min(Math.ceil(fwhm * factor), Math.pow(2, 25) - 1);
        if (length % 2 === 0)
            length++;
    }
    const center = (length - 1) / 2;
    const data = new Float64Array(length);
    for (let i = 0; i <= center; i++) {
        data[i] = fct(i - center, fwhm, mu) * height;
        data[length - 1 - i] = data[i];
    }
    return data;
}
exports.getData = getData;
